README.md
# progress_loading_button

![GitHub repo size](https://img.shields.io/github/repo-size/gairick-saha/progress_loading_button.svg)
![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/gairick-saha/progress_loading_button.svg)
![GitHub top language](https://img.shields.io/github/languages/top/gairick-saha/progress_loading_button.svg)
[![GitHub issues](https://img.shields.io/github/issues/gairick-saha/progress_loading_button.svg)](https://github.com/gairick-saha/progress_loading_button/issues)
[![GitHub license](https://img.shields.io/github/license/gairick-saha/progress_loading_button.svg)](https://github.com/gairick-saha/progress_loading_button/blob/master/LICENSE)

**progress_loading_button** is a free and open source (MIT license) Material Flutter Button that supports variety of buttons style demands. It is designed to be easy to use and customizable.

## Get started

Add this to your package's pubspec.yaml file:

```yaml
progress_loading_button: '^1.0.0'
```

### **Install it**

You can install packages from the command line:

```
$ flutter pub get
```

Alternatively, your editor might support flutter pub get.

### **Import it**

Now in your Dart code, you can use:

```dart
import 'package:progress_loading_button/progress_loading_button.dart';

```

## How to use

Add `LoadingButton` to your widget tree:

```dart
LoadingButton(
    defaultWidget: Text('Click Me'),
    width: 196,
    height: 60,
    onPressed: () async {
        await Future.delayed(
            Duration(milliseconds: 3000),
            () {
                print('Button Pressed');
            },
        );
    },
)
```

More parameters:

```dart
LoadingButton({
    Key? key,
    required this.defaultWidget,
    this.loadingWidget = const CircularProgressIndicator(),
    required this.onPressed,
    this.type = LoadingButtonType.Raised,
    this.color,
    this.width = double.infinity,
    this.height = 40.0,
    this.borderRadius = 5.0,
    this.borderSide = BorderSide.none,
    this.animate = true,
  }) : super(key: key);
```

Three types supported:

```dart
enum LoadingButtonType {
  Raised,
  Flat,
  Outline,
}
```

## Example

![](./demo/progress_loading_button.gif)

## Source

Source code and example of this library can be found in git:

```
$ git clone https://github.com/gairick-saha/progress_loading_button.git
```


CHANGELOG.md
## 1.1.1

* First release

## 1.1.2

* Demo Screenshot added

## 1.1.3

* Minor bug fixes

## 1.1.4

* Minor bug fixes

## 1.1.5

* LoadingButtonTypes renamed from Raised to Elevated and Flat to Text
* loadingWidget is now nullable by default it will show default loading widget

## 1.1.6

* Minor bug fixes

## 1.1.7

* Minor bug fixes

## 1.1.8

* Minor bug fixes

## 1.1.9

* Minor bug fixes

## 2.0.0

* Minor bug fixes

## 2.0.1

* Minor bug fixes

lib/src/loading_button.dart
import 'dart:io';

import 'package:flutter/cupertino.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

enum LoadingButtonState {
  Default,
  Processing,
}

enum LoadingButtonType {
  Elevated,
  Text,
  Outline,
}

class LoadingButton extends StatefulWidget {
  const LoadingButton({
    Key? key,
    required this.defaultWidget,
    this.loadingWidget,
    required this.onPressed,
    this.type = LoadingButtonType.Elevated,
    this.color,
    this.textcolor,
    this.width,
    this.height = kMinInteractiveDimension,
    this.borderRadius = 5.0,
    this.borderSide = BorderSide.none,
    this.animate = true,
    this.padding,
  }) : super(key: key);

  final Widget defaultWidget;
  final Widget? loadingWidget;
  final Future Function()? onPressed;
  final LoadingButtonType? type;
  final Color? color;
  final Color? textcolor;
  final double? width;
  final double? height;
  final double? borderRadius;
  final BorderSide? borderSide;
  final bool animate;
  final EdgeInsetsGeometry? padding;

  @override
  State<LoadingButton> createState() => _LoadingButtonState();
}

class _LoadingButtonState extends State<LoadingButton>
    with TickerProviderStateMixin {
  final GlobalKey _globalKey = GlobalKey();
  Animation? _anim;
  AnimationController? _animController;
  final Duration _duration = const Duration(milliseconds: 250);
  LoadingButtonState? _state;
  double? _width;
  double? _height;
  double? _borderRadius;
  BorderSide? _borderSide;

  @override
  void dispose() {
    _reset(isDisposed: true);
    _animController?.dispose();
    super.dispose();
  }

  @override
  void initState() {
    _reset();
    super.initState();
  }

  @override
  void didUpdateWidget(covariant LoadingButton oldWidget) {
    if (oldWidget.width != widget.width ||
        oldWidget.height != widget.height ||
        oldWidget.borderRadius != widget.borderRadius ||
        oldWidget.borderSide != widget.borderSide) {
      _reset();
    }
    super.didUpdateWidget(oldWidget);
  }

  void _reset({bool isDisposed = false}) {
    if (mounted) {
      _state = LoadingButtonState.Default;
      _width = widget.width;
      _height = widget.height;
      _borderRadius = widget.borderRadius;
      _borderSide = widget.borderSide;
      if (!isDisposed) {
        setState(() {});
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      key: _globalKey,
      height: _height,
      width: _width,
      child: _buildChild(context),
    );
  }

  Widget _buildChild(BuildContext context) {
    final ButtonStyle textbuttonStyle = TextButton.styleFrom(
      foregroundColor: widget.textcolor,
      disabledForegroundColor: widget.textcolor,
      padding: _state == LoadingButtonState.Processing
          ? EdgeInsets.zero
          : widget.padding ??
              ButtonStyleButton.scaledPadding(
                const EdgeInsets.symmetric(horizontal: 16),
                const EdgeInsets.symmetric(horizontal: 8),
                const EdgeInsets.symmetric(horizontal: 4),
                MediaQuery.maybeOf(context)?.textScaleFactor ?? 1,
              ),
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(_borderRadius!),
        side: _borderSide!,
      ),
      splashFactory: NoSplash.splashFactory,
      alignment: Alignment.center,
    );
    final ButtonStyle elevatedbuttonStyle = ElevatedButton.styleFrom(
      backgroundColor: widget.color,
      disabledBackgroundColor:
          widget.color ?? Theme.of(context).colorScheme.primary,
      foregroundColor: widget.textcolor,
      disabledForegroundColor:
          widget.textcolor ?? Theme.of(context).colorScheme.surface,
      elevation: _state! == LoadingButtonState.Processing ? 0 : null,
      padding: _state == LoadingButtonState.Processing
          ? EdgeInsets.zero
          : widget.padding ??
              ButtonStyleButton.scaledPadding(
                const EdgeInsets.symmetric(horizontal: 16),
                const EdgeInsets.symmetric(horizontal: 8),
                const EdgeInsets.symmetric(horizontal: 4),
                MediaQuery.maybeOf(context)?.textScaleFactor ?? 1,
              ),
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(_borderRadius!),
        side: _borderSide!,
      ),
      splashFactory: NoSplash.splashFactory,
      alignment: Alignment.center,
    );
    final ButtonStyle outlinedbuttonStyle = OutlinedButton.styleFrom(
      side: _borderSide ??
          BorderSide(
            color: widget.color ?? Theme.of(context).colorScheme.primary,
          ),
      foregroundColor: widget.textcolor ?? widget.color,
      disabledForegroundColor: widget.textcolor ?? widget.color,
      padding: _state == LoadingButtonState.Processing
          ? EdgeInsets.zero
          : widget.padding ??
              ButtonStyleButton.scaledPadding(
                const EdgeInsets.symmetric(horizontal: 16),
                const EdgeInsets.symmetric(horizontal: 8),
                const EdgeInsets.symmetric(horizontal: 4),
                MediaQuery.maybeOf(context)?.textScaleFactor ?? 1,
              ),
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(_borderRadius!),
        side: _borderSide!,
      ),
      splashFactory: NoSplash.splashFactory,
      alignment: Alignment.center,
    );

    switch (widget.type!) {
      case LoadingButtonType.Elevated:
        return ElevatedButton(
          style: elevatedbuttonStyle,
          onPressed: _onButtonPressed(),
          child: _buildButtonChild(
            context,
            loadingColor:
                widget.textcolor ?? Theme.of(context).colorScheme.surface,
          ),
        );
      case LoadingButtonType.Text:
        return TextButton(
          style: textbuttonStyle,
          onPressed: _onButtonPressed(),
          child: _buildButtonChild(
            context,
            loadingColor: widget.color ?? Theme.of(context).colorScheme.primary,
          ),
        );
      case LoadingButtonType.Outline:
        return OutlinedButton(
          style: outlinedbuttonStyle,
          onPressed: _onButtonPressed(),
          child: _buildButtonChild(
            context,
            loadingColor: widget.color ?? Theme.of(context).colorScheme.primary,
          ),
        );
    }
  }

  Widget _buildButtonChild(BuildContext context,
      {required Color loadingColor}) {
    Widget buttonWidget;
    switch (_state!) {
      case LoadingButtonState.Default:
        buttonWidget = widget.defaultWidget;
        break;
      case LoadingButtonState.Processing:
        buttonWidget = _buildLoadingIndicator(context, color: loadingColor);
        break;
    }
    return buttonWidget;
  }

  Widget _buildLoadingIndicator(BuildContext context, {required Color color}) {
    return widget.loadingWidget ??
        (!kIsWeb && Platform.isIOS
            ? CupertinoActivityIndicator(
                color: color,
              )
            : CircularProgressIndicator(
                color: color,
                backgroundColor: Colors.transparent,
              ));
  }

  VoidCallback? _onButtonPressed() {
    return _state == LoadingButtonState.Processing
        ? null
        : () async {
            if (_state != LoadingButtonState.Default) {
              return;
            }

            VoidCallback? onDefault;
            if (widget.animate) {
              _toProcessing();
              _forward((status) {
                if (status == AnimationStatus.dismissed) {
                  _toDefault();
                  if (onDefault != null) {
                    onDefault();
                  }
                }
              });
              onDefault =
                  widget.onPressed == null ? null : await widget.onPressed!();
              _reverse();
            } else {
              _toProcessing();
              onDefault =
                  widget.onPressed == null ? null : await widget.onPressed!();
              _toDefault();
              if (onDefault != null) {
                onDefault();
              }
            }
          };
  }

  void _toProcessing() => setState(() {
        _state = LoadingButtonState.Processing;
      });

  void _toDefault() {
    if (mounted) {
      setState(() {
        _state = LoadingButtonState.Default;
      });
    } else {
      _state = LoadingButtonState.Default;
    }
  }

  void _forward(AnimationStatusListener stateListener) {
    double initialWidth = _globalKey.currentContext!.size!.width;
    double initialBorderRadius = widget.borderRadius!;
    double targetWidth = _height!;
    double targetBorderRadius = _height! / 2;

    _animController = AnimationController(duration: _duration, vsync: this);
    _anim = Tween(begin: 0.0, end: 1.0).animate(_animController!)
      ..addListener(() {
        setState(() {
          _width = initialWidth - ((initialWidth - targetWidth) * _anim!.value);
          _borderRadius = initialBorderRadius -
              ((initialBorderRadius - targetBorderRadius) * _anim!.value);
        });
      })
      ..addStatusListener(stateListener);

    _animController?.forward();
  }

  void _reverse() {
    _animController?.reverse();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:progress_loading_button/progress_loading_button.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        visualDensity: VisualDensity.adaptivePlatformDensity,
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          centerTitle: true,
          title: Text('Progress Loading Button Example'),
        ),
        body: Center(
          child: LoadingButton(
            defaultWidget: const Text('Click Me'),
            width: 196,
            height: 60,
            onPressed: () async {
              await Future.delayed(
                const Duration(milliseconds: 3000),
                () {
                  print('Button Pressed');
                },
              );
            },
          ),
        ),
      ),
    );
  }
}
